rendermodules/rough_align/apply_rough_alignment_to_montages.py

```diff
@@ -61,39 +66,157 @@
+                results[t.tileId] = t.ip[0].maskUrl
+
+    elif mask_input_dir is not None:
+        tids = [t.tileId for t in tilespecs]
+        maskfiles = []
+        for ext in exts:
+            e = mask_input_dir + '/*.' + ext
+            maskfiles += glob.glob(e)
+        maskbasenames = np.array([os.path.basename(os.path.splitext(p)[0])
+                                  for p in maskfiles])
+        for t in tids:
+            ind = np.argwhere(maskbasenames == t).flatten()
+            if ind.size > 1:
+                raise RenderModuleException("more than one mask in "
+                                            "directory %s matches tileId %s" %
+                                            (mask_input_dir, t))
+            if ind.size == 0:
+                continue
+            results[t] = pathlib.Path(maskfiles[ind[0]]).as_uri()
+
+    return results
+
+
+def add_masks_to_lowres(render, stack, z, mask_map):
+    if len(mask_map) == 0:
+        return
+
+    tilespecs = renderapi.tilespec.get_tile_specs_from_z(
+            stack, z, render=render)
+
+    for t in tilespecs:
+        if t.tileId in mask_map:
+            t.ip['0'].maskUrl = mask_map[t.tileId]
+            renderapi.client.import_tilespecs(
+                    stack,
+                    [t],
+                    render=render)
+    return
+
+
+def filter_highres_with_masks(resolved_highres, tspec_lowres, mask_map):
+    """function to return a filtered list of tilespecs from a
+       ResolvedTiles object, based on a lowres mask
+
+    Parameters
+    ----------
+    resolved_highres : renderapi.resolvedtiles.ResolvedTiles object
+        tilespecs and transforms from a single section
+    tspec_lowres: renderapi.tilespec.TileSpec object
+        tilespec from a downsampled stack
+    mask_map: dict
+        keys should match lowres tileids, values are mask file URI
+
+    Returns
+    -------
+    new_highres: List of renderapi.tilespec.TileSpec objects
+        which highres specs are fully contained
+        within the boundary of mask=255
+    """
+
+    if tspec_lowres.tileId not in mask_map.keys():
+        return resolved_highres.tilespecs
+
+    impath = urllib.parse.unquote(
+                 urllib.parse.urlparse(
+                     mask_map[tspec_lowres.tileId]).path)
+    maskim = cv2.imread(impath, 0)
+    maskim = 255 - maskim
+    _, contours, _ = cv2.findContours(
+                         maskim,
+                         cv2.RETR_TREE,
+                         cv2.CHAIN_APPROX_SIMPLE)
+
+    mask_polygons = []
+    for c in contours:
+        c = c.squeeze().astype('float')
+        for tf in tspec_lowres.tforms:
+            c = tf.tform(c)
+        mask_polygons.append(Polygon(c).buffer(0))
+
+    new_highres = []
+    for t in resolved_highres.tilespecs:
+        tc = t.bbox_transformed(
+                    reference_tforms=resolved_highres.transforms)
+        tpoly = Polygon(tc).buffer(0)
+        for p in mask_polygons:
```

Comment:
Will this result in duplicate tilespecs if there are multiple mask polygons?  I am not sure if this causes a problem.

Reply:
In my experience, this does not cause a problem. In the importjson render client, a dictionary is created which enforces unique keys, so duplicates don't come through.
https://github.com/saalfeldlab/render/blob/master/render-app/src/main/java/org/janelia/alignment/spec/ResolvedTileSpecCollection.java

you can see it here:
```
In [93]: resolved = renderapi.resolvedtiles.get_resolved_tiles_from_z('FUSEDOUTSTACK_Chunk_31_33', 21997, render=render)

In [94]: tspec = resolved.tilespecs[0]

In [95]: tspecs = [tspec, tspec, tspec]

In [96]: renderapi.stack.create_stack('test_can_delete', render=render)
Out[96]: <Response [201]>

In [97]: renderapi.stack.set_stack_state('test_can_delete', state='LOADING', render=render)
Out[97]: <Response [201]>

In [98]: renderapi.client.import_tilespecs('test_can_delete', tspecs, sharedTransforms=resolved.transforms, render=render)
org.janelia.render.client.ImportJsonClient

  Running: /usr//bin/java -cp /allen/aibs/pipeline/image_processing/volume_assembly/render-jars/production/render-ws-java-client-standalone.jar -Xms1G -Xmx1G -Djava.awt.headless=true -XX:+UseSerialGC org.janelia.render.client.ImportJsonClient --baseDataUrl http://em-131fs:8987/render-ws/v1 --owner TEM --project 17797_1R --stack test_can_delete --transformFile /tmp/tmpntcV2I.json /tmp/tmp7sjENv.json


04:02:00.577 [main] INFO  [org.janelia.render.client.ClientRunner] run: entry
04:02:00.686 [main] INFO  [org.janelia.render.client.ImportJsonClient] runClient: entry, parameters={
  "renderWeb" : {
    "baseDataUrl" : "http://em-131fs:8987/render-ws/v1",
    "owner" : "TEM",
    "project" : "17797_1R"
  },
  "tileSpecValidator" : { },
  "stack" : "test_can_delete",
  "transformFile" : "/tmp/tmpntcV2I.json",
  "tileFiles" : [
    "/tmp/tmp7sjENv.json"
  ]
}
04:02:00.845 [main] INFO  [org.janelia.render.client.RenderDataClient] getStackMetaData: submitting GET http://em-131fs:8987/render-ws/v1/owner/TEM/project/17797_1R/stack/test_can_delete
04:02:00.911 [main] INFO  [org.janelia.render.client.RenderDataClient] ensureStackIsInLoadingState: test_can_delete stack is already in the LOADING state
04:02:00.911 [main] INFO  [org.janelia.render.client.ImportJsonClient] loadTransformData: entry, path=/tmp/tmpntcV2I.json
04:02:00.961 [main] INFO  [org.janelia.render.client.ImportJsonClient] loadTransformData: exit, loaded 1 transform specs
04:02:00.961 [main] INFO  [org.janelia.render.client.ImportJsonClient] importStackData: entry, tileFile=/tmp/tmp7sjENv.json
04:02:00.961 [main] INFO  [org.janelia.render.client.ImportJsonClient] loadTileData: entry, path=/tmp/tmp7sjENv.json
04:02:00.978 [main] INFO  [org.janelia.render.client.ImportJsonClient] loadTileData: exit, loaded 3 tile specs
04:02:01.069 [main] INFO  [org.janelia.render.client.ImportJsonClient] importStackData: derived bounding box for 1 tiles, elapsedSeconds=0
04:02:01.093 [main] INFO  [org.janelia.render.client.RenderDataClient] saveResolvedTiles: submitting PUT http://em-131fs:8987/render-ws/v1/owner/TEM/project/17797_1R/stack/test_can_delete/resolvedTiles for 1 transforms and 1 tiles
04:02:01.754 [main] INFO  [org.janelia.render.client.ImportJsonClient] importStackData: exit, saved tiles and transforms from /tmp/tmp7sjENv.json
04:02:01.755 [main] INFO  [org.janelia.render.client.ClientRunner] run: exit, processing completed in 0 hours, 0 minutes, 1 seconds

In [99]: renderapi.stack.set_stack_state('test_can_delete', state='COMPLETE', render=render)
Out[99]: <Response [201]>
```